DM-28494: Allow DBB endpoint manager to skip ingesting files based upon filename pattern

mxk62 opened the pull request.

mxk62 committed:
Simplify communication between threads

The provide some degree of scalability the Ingester is able to use a
pool of ingest threads.  However, the communication between the main
thread and the ingest threads was somewhat lacking:

* The input queue required messages to be just paths to files.  As
  a result, passing any other useful pieces of information about a
  file was not possible.  As a result, maintaining auxiliary data
  structures was necessary to compile a valid database record from an
  input message and the corresponding output message.

* There were two output queues which were only used to determine
  whether an ingest attempt was a success or a failure.

To address these shortcomings, I introduced two new data structures
representing the input and the output messages:

1. a request for an ingest attempt which holds any piece of
   information about the file for which the ingest attempt is
   requested,

2. a reply which captures outcome of a ingest attempt and any other data
   required to create a corresponding database entry.

Such an approach no longer requires maintaining the auxiliary data
structures and reduced number of message queues to two.

mxk62 committed:
Improve docstrings and comments

mxk62 committed:
Allow ingester to be more selective about files

Ingester makes ingest attempts for all files that are tracked and
had no mechanism to exclude certain files which should not be ingested
to a Butler repository.  To address this shortcoming and make tracking
files easier I implemented a filtering mechanism which will allow an
operator to ignore select files based on their filenames.

mxk62 committed:
Make naming consistent among components

MichelleGower commented on etc/gen2ingester.yaml:
```diff
@@ -24,6 +24,8 @@ ingester:
       create: false
       ignoreIngested: false
 
+  includelist: null
```
Missing new documentation (get-started.rst) for the new includelist, excludelist in ingester?

mxk62 replied:
`get-started.rst` is a quick start guide type of the document. It is meant to let a person new to DBB endpoint buffer manager get it working with minimal effort. So I intentionally do not cover all bells and whistles in it, just the basics.

If you think that the filtering mechanism is important enough to be covered in this type of document, I could "sneak in" a paragraph or two. Alternatively, I can write a separate document describing all currently available configuration options. But I'd prefer to do that work in a separate ticket.

MichelleGower replied:
It was just confusing that it is in the get-started.rst for a different piece (e.g., backfill tool).

mxk62 replied:
Yeah, that's an error on my part. I copy-pasted the actual config I used in tests and forgot to edit it out. Will fix.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -126,95 +176,110 @@ def run(self):
+
+                message, status = None, None
+                if not any(re.search(patt, path) for patt in self.includelist):
+                    message = "doesn't match search criteria: not allowed"
```
Perhaps reword message to something like: doesn't match search includelist criteria   "Not allowed" is technically true for exclusion patterns too.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -126,95 +176,110 @@ def run(self):
+                    message = "doesn't match search criteria: not allowed"
+                    status = Status.IGNORED
+                elif any(re.search(patt, path) for patt in self.excludelist):
+                    message = "doesn't match search criteria: excluded"
```
It will be helpful for debugging if we could store the pattern which caused the exclusion.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:
```diff
@@ -99,7 +99,7 @@ def __init__(self, config):
 
         search = config["search"]
-        self.search_opts = dict(blacklist=search.get("blacklist", None))
+        self.search_opts = dict(excludelist=search.get("excludelist", None))
```
excludelist and includelist are normally two separate words...they would read easier if had underscore/capital letter

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -39,8 +41,54 @@
+    """
+
+    status: Status = None
+    """Status of the ingest attempt
```
All the other comments end with a period so happened to notice that this one didn't.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -126,95 +176,110 @@ def run(self):
                     continue
-                inp.put(path)
+
+                req = Request(
```
Add a comment above this section (e.g., file passed search criteria, so create a request to ingest it)

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -126,95 +176,110 @@ def run(self):
-        Queries the database for newly discovered files and enqueues them for
-        processing.
+        Queries the database for new files.  A file is considered new if
+        there is no events associated with it.
```
There are no events

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -126,95 +176,110 @@ def run(self):
+        For each selected file, a new event is recorded in the database with
+        the status set to UNTRIED.
+
+        No events will be created if any errors are encountered.
```
Any errors in/during what?  Or maybe "No UNTRIED events".  I know this is in a particular function, but the first time I read this I immediately thought of ingestion failure events which we do want (but then figured out that they aren't in the scope of this function).

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -126,95 +176,110 @@ def run(self):
+
+        No events will be created if any errors are encountered.
         """
         # Find new files that is files for which there is no recorded events:
```
is -> are:  there are no recorded events

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -273,53 +348,26 @@ def _grab(self):
+        records = []
         try:
-            for rec in query:
+            records = list(query)
```
I don't think this change to a pre-fetch all into a list is an ideal change if the query results are large (i.e., any db fetching in batches, for speed and reducing memory, is lost).  I doubt we're currently at the size where this matters, but thought I would mention it.

mxk62 replied:
I see your point, but in this case the number of the results this query can return is always capped by `batch_size` parameter. This parameter should be a fairly small positive integer (default is 10).

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -273,53 +348,26 @@ def _grab(self):
+                self.session.commit()
+            except SQLAlchemyError as ex:
+                self.session.rollback()
+                logger.error(f"cannot commit updates: {ex}")
```
Should records be reset to empty list if have a commit error (based upon docstring)?

mxk62 replied:
It definitely should. Good catch!

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/search.py:
```diff
@@ -57,17 +57,17 @@ def register(gen):
-    Generates the file paths in a given directory tree excluding those files
-    that were blacklisted.
+    Generates the file paths in a given directory tree ignoring files
+    that names match select patterns.
```
files that names -> either "files that have names matching" or "files with names that match" or "files whose names"

mxk62 committed:
Let the ingester handle ingestion issues

Any issues with image ingestion were handled on the level of ingestion
plugins.  To reduce the duplicated code I let the ingester handle these
issues instead.

mxk62 committed:
Add a support for new LSST ingest option

Recently a new ingest option was added to LSST ingest software allowing
the ingest process fail early if any issues were encountered.  Added
support for it in Gen3 ingest plugin.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -328,12 +376,9 @@ def worker(inp, out, err, plugin_cls=None, plugin_cfg=None):
+        Channel for gathering the results of the ingest requests.
     plugin_cls : `Plugin`, optional
         An ingest plugin to use. If None (default), nothing will be
         done, effectively a no-op.
```
Line not in current changes, but after just thinking about the return value on a previous method this stood out.  By "nothing will be done" does that mean nothing inp and out stay the same?

mxk62 replied:
No, the worker will keep processing requests from the input channel and putting replies in the output one. Only the ingest attempt will be a NO-OP.

Realistically speaking, these defaults can be used only for testing inter-thread communication, not during normal operation. So I guess they're not good defaults after all. I will make `plugin_cls` and `plugin_cfg` mandatory arguments.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/ingester/ingester.py:
```diff
@@ -352,25 +397,31 @@ def worker(inp, out, err, plugin_cls=None, plugin_cfg=None):
+            exc_msg = traceback.format_exception_only(type(exc), exc)[0]
+            message = exc_msg.strip()
+            status = Status.FAILURE
+            logger.error(message)
```
The code is now catching a wide range of errors (including Syntax).  Not saving the full traceback into the database was part of the design.  Does all of the traceback go into a log that could be manual checked later?

mxk62 replied:
> The code is now catching a wide range of errors (including Syntax)

I'll make the manager handle `SyntaxError` exception as a special case.

> Does all of the traceback go into a log that could be manual checked later?

I'll make sure they will be logged.

MichelleGower commented on python/lsst/dbb/buffmngrs/endpoint/search.py:
```diff
@@ -79,18 +79,18 @@ def scan(directory, blacklist=None, **kwargs):
         for fn in filenames:
             path = os.path.relpath(os.path.join(dirpath, fn), start=directory)
-            if any(re.search(patt, path) for patt in blacklist):
+            if any(re.search(patt, path) for patt in excludelist):
```
Since I commented on the ingestion ignoring logging, I noticed in both of these search methods, there isn't a logging message when ignoring files (not even at the DEBUG logging level).  Consider if should add it.  Since have manual ways of checking for ignored files, can merge without it if want to wait to for an actual use case/requirement.

mxk62 replied:
I'll make the manager to log DEBUG messages when files are ignored.

MichelleGower reviewed (approved):
Mostly minor changes ranging from comment wording, add comments, to a couple of logging change requests.

mxk62 committed:
Get rid of defaults for ingest plugin worker

The worker was providing the defaults for the ingests plugin
specification.  Realistically speaking, these defaults could only be
useful during testing/debugging.  Hence they were not good defaults so
made these arguments mandatory, not optional.

mxk62 committed:
Make logging messages more informative

Included some additional piece of information to logging messages, e.g.,
patterns from include and exclude lists that match a file name and make
sure that trace stacks which occurred during ingestions attempt will be
logged too.

mxk62 committed:
Trim down options in example config

mxk62 committed:
Use snake case in exclude and include lists
